Allow null or undefined in Reference Records

codehag opened the pull request:
For review by @jorendorff.

I checked all locations where `.[[Base]]` is used. There don't appear to be locations as far as I can tell, where this change will be an issue, see https://searchfox.org/ecma262/search?q=.%5B%5BBase%5D%5D&path=

In this cases, null and undefined is handled implicitly: [IsUnresolvableReference](https://searchfox.org/ecma262/rev/6d6b73ab6fe3b070737a5cbded19337a8b9e4c9c/spec.html#3808) and this change wouldn't result in a change in behavior.

The same is true of [IsPropertyReference](https://searchfox.org/ecma262/rev/6d6b73ab6fe3b070737a5cbded19337a8b9e4c9c/spec.html#3799), but I think we want to change it, because otherwise we would fall back on thinking this is an environment record. See the code around these [instances](https://searchfox.org/ecma262/search?q=IsPropertyReference&redirect=false). But, if we change to return true for null or undefined, we will be caught by `ToObject` as in [these cases](https://searchfox.org/ecma262/search?q=IsPropertyReference.*+%5C*true&path=&case=false&regexp=true).

I am not as sure about [`ToObject`](https://searchfox.org/ecma262/search?q=ToObject%28_V_.%5B%5BBase%5D%5D&path=&case=false&regexp=false) -- this will change at which point we throw (which is, I guess, the point). 

In all of these cases, we follow with [an assertion that the value is an environment record](https://searchfox.org/ecma262/search?q=be+_.*_%5C.%5C%5B%5C%5BBase%5C%5D%5C%5D&path=&case=false&regexp=true)

I may have missed something that needs to be changed though, let me know if that is the case.

jmdyck committed:
Normative: Make B.1.1 "Numeric Literals" normative (#1867)

(Part of Annex B reform, see PR #1595.)

jmdyck committed:
Normative: Make B.1.2 "String Literals" normative (#1867)

(Part of Annex B reform, see PR #1595.)

B.1.2 makes 2 changes to the EscapeSequence production:
(1) It adds the rhs `NonOctalDecimalEscapeSequence`.
(2) It replaces the rhs:
        `0` [lookahead &lt;! DecimalDigit]
    with:
        LegacyOctalEscapeSequence
    where the latter nonterminal generates `0` among lots of other things.

Change 1 is straightforward, but change 2 is tricky.
In the EscapeSequence production, we can't simply replace
the `0` alternative with LegacyOctalEscapeSequence (as B.1.2 does),
because the `0` alternative must be treated differently
from everything else that LegacyOctalEscapeSequence derives.
(The `0` alternative is allowed in contexts where
everything else that LegacyOctalEscapeSequence derives is forbidden.)
So instead, we redefine LegacyOctalEscapeSequence to exclude the `0` alternative.
Specifically, the 'overlap' comes from:

    LegacyOctalEscapeSequence ::
        OctalDigit [lookahead &notin; OctalDigit]

so we replace that with:

    LegacyOctalEscapeSequence ::
        `0` [lookahead &isin; {`8`, `9`}]
        NonZeroOctalDigit [lookahead &notin; OctalDigit]

(See Issue #1975 for more details.)
Resolves #1975.

jorendorff commented on spec.html:
```diff
@@ -13567,7 +13567,6 @@ <h1>EvaluatePropertyAccessWithExpressionKey ( _baseValue_, _expression_, _strict
         1. Let _propertyNameValue_ be ? GetValue(_propertyNameReference_).
-        1. Let _bv_ be ? RequireObjectCoercible(_baseValue_).
         1. Let _propertyKey_ be ? ToPropertyKey(_propertyNameValue_).
         1. Return the Reference Record { [[Base]]: _bv_, [[ReferencedName]]: _propertyKey_, [[Strict]]: _strict_, [[ThisValue]]: ~empty~ }.
```
Since _bv_ has been deleted, the use of `_bv_` here must be changed to `_baseValue_`.

jorendorff commented on spec.html:
```diff
@@ -13577,7 +13576,6 @@ <h1>EvaluatePropertyAccessWithIdentifierKey ( _baseValue_, _identifierName_, _st
         1. Assert: _identifierName_ is an |IdentifierName|.
-        1. Let _bv_ be ? RequireObjectCoercible(_baseValue_).
         1. Let _propertyNameString_ be StringValue of _identifierName_.
         1. Return the Reference Record { [[Base]]: _bv_, [[ReferencedName]]: _propertyNameString_, [[Strict]]: _strict_, [[ThisValue]]: ~empty~ }.
```
And here.

jorendorff reviewed (commented):
* [x] I can't tell if MakeSuperPropertyReference needs the same treatment. Maybe someone more familiar with the `super` bits of the spec can say.

* [x] I gather the intent is that property expressions produce property References and identifiers produce environment References—it's a static invariant.

    I think the language of the check in IsPropertyReference should be changed to:

    > 2.  If *V*.[[Base]] is an Environment Record, return **false**; otherwise return **true**.

* [x] In [GetValue](https://tc39.es/ecma262/#sec-getvalue), the call to "! ToObject" must be changed to "? ToObject" since it can now fail.

* [x] Same thing in [PutValue](https://tc39.es/ecma262/#sec-putvalue).

* [x] Same thing in [Evaluation for `delete` expressions](https://tc39.es/ecma262/#sec-delete-operator-runtime-semantics-evaluation), step 5.b.

jorendorff commented:
OK, I think MakeSuperPropertyReference needs the same change, for consistency.

But there is something I'm confused about.

There are actually *two* cases where the RequireObjectCoercible check in MakeSuperPropertyReference could fail:

- One is if *env*.GetSuperBase() returns **null** in step 3. I think this is what happens in this example:

    ```
    class C {
      set v(x) {
        super.v = console.log("GOT HERE");
      }
    }
    new C().v = 37;
    ```

- The other is if *env*.GetSuperBase() returns **undefined**. I don't know how this is possible.

codehag commented:
> The other is if env.GetSuperBase() returns undefined. I don't know how this is possible.

I don't know if it is? In the code above `env.GetSuperBase()` we assert that it is not undefined, so I am not sure how we can get to that. Maybe this is some older text that made more sense before? I will see if I can get more eyes on it. I made the change in any case.

leobalter committed:
Editorial: Add ECMA-402 as a Downstream dependency (#2268)

bakkot committed:
Editorial: reword StringIndexOf in terms of substring (#2258)

Jack-Works committed:
Editorial: Add Yield, CreateIteratorFromClosure, CreateAsyncIteratorFromClosure abstract ops (#2045)

ryzokuken committed:
Editorial: move Section 26.6 Promise Objects (#1234)

Move Section 26.6 Promise Objects up to become Section 26.2, below
Section 26.1 Iteration and Section 26.3 GeneratorFunction Objects,
thus also moving Sections 26.5 AsyncGenerator Objects and 26.7
AsyncFunction Objects together.

Fixes #1170

bakkot committed:
Markup: write URI pattern without grammarkdown notation (#2270)

bakkot committed:
Markup: fix links to LegacyOctalIntegerLiteral and NonOctalDecimalIntegerLiteral in Annex C (#2270)

bakkot committed:
Markup: use bare "example" attribute rather than "type=example" (#2270)

bakkot committed:
Editorial: use `/`, not `&divide;` (#2270)

bakkot committed:
Meta: bump ecmarkup to 5.0.0 (#2270)

bakkot committed:
Editorial: make EPSILON unambiguously positive (#2272)

jmdyck committed:
Editorial: Delete "the Number value for" in Number::leftShift (#2273)

For consistency with Number::signedRightShift and Number::unsignedRightShift, the result of a left or right shift is a Number, so it doesn't make sense to get "the Number value for" that result.

caiolima commented:
I spent some time investigating the impact of such change and couldn't find anything strange on `super.prop` cases, nor `delete` or `typeof` operators.

> > The other is if env.GetSuperBase() returns undefined. I don't know how this is possible.
> 
> I don't know if it is? In the code above `env.GetSuperBase()` we assert that it is not undefined, so I am not sure how we can get to that. Maybe this is some older text that made more sense before? I will see if I can get more eyes on it. I made the change in any case.

Looking into [MakeSuperPropertyReference](https://tc39.es/ecma262/#sec-makesuperpropertyreference) we can have a clue that if we ever get `env.GetSuperBase()` to return `undefined`, it's a bug, since `env.HasSuperBinding()` would return `false` when we get `[[HomeObject]]` to be `undefined`. I'm also not able to think in a way where we could have `super.prop`, where `super` evaluates to `undefined`.

As a matter of clarification, I managed to make super evaluate to `null` with this code:

```js
class C {
    static access() {
        super.foo = print('blah');
    }
}

Object.setPrototypeOf(C, null);
C.access();
```

Regarding `typeof` operation, since it relies on `GetValue(V)` and the PR includes changes to it, we should not have troubles there.
Overall, I think such change makes sense and LGTM.

bakkot committed:
Editorial: split out ForDeclarationBindingInitialization from BindingInitialization (#2271)

bakkot committed:
Editorial: avoid overloading StringValue for RegExp capturing group identifiers (#2271)

bakkot committed:
Editorial: consolidate PropertyDefinitionEvaluation (#2271)

bakkot committed:
Editorial: split out MethodDefinitionEvaluation from PropertyDefinitionEvaluation (#2271)

bakkot committed:
Editorial: consolidate LabelledEvaluation (#2271)

bakkot committed:
Editorial: split out LoopEvaluation from LabelledEvaluation (#2271)

bakkot committed:
Editorial: use SDO-style dispatch in LabelledEvaluation, rather than switching in the algorithm (#2271)

bakkot committed:
Editorial: factor out parameter productions into their own section (#2271)

bakkot committed:
Editorial: consolidate ContainsExpression, IsSimpleParameterList, HasInitializer, and ExpectedArgumentCount (#2271)

bakkot committed:
Editorial: split ContainsUseStrict into FunctionBody, ConciseBody, and AsyncConciseBody cases (#2271)

bakkot committed:
Editorial: introduce section for syntax-directed operations (#2271)

bakkot committed:
Editorial: consolidate scope-analysis related SDOs (#2271)

bakkot committed:
Editorial: consolidate notes about *default* in BoundNames (#2271)

bakkot committed:
Editorial: split out the different kinds of body for EvaluateBody (#2271)

bakkot committed:
Editorial: split out the different kinds of declaration for InstantiateFunctionObject (#2271)

bakkot committed:
Editorial: invoke SV on StringLiteral directly rather than through StringValue (#2271)

bakkot committed:
Editorial: consolidate StringValue (#2271)

bakkot committed:
Editorial: consolidate label-related SDOs (#2271)

bakkot committed:
Editorial: consolidate BindingInitialization (#2271)

bakkot committed:
Editorial: consolidate IteratorBindingInitialization (#2271)

bakkot committed:
Editorial: consolidate AssignmentTargetType (#2271)

bakkot committed:
Editorial: delete case of AssignmentTargetType covered by chain rule (#2271)

bakkot committed:
Editorial: consolidate PropName (#2271)

bakkot committed:
Editorial: consolidate HasName (#2271)

bakkot committed:
Editorial: delete case of IsFunctionDefinition covered by chain rule (#2271)

bakkot committed:
Editorial: consolidate ArgumentListEvaluation (#2271)

bakkot committed:
Editorial: consolidate IsDestructuring (#2271)

bakkot committed:
Editorial: consolidate HasDirectSuper (#2271)

bakkot committed:
Editorial: consolidate ExportedBindings (#2271)

bakkot committed:
Editorial: consolidate ExportedNames (#2271)

bakkot committed:
Editorial: consolidate ExportEntries (#2271)

bakkot committed:
Editorial: consolidate ImportEntries (#2271)

bakkot committed:
Editorial: consolidate ModuleRequests (#2271)

bakkot committed:
Editorial: give AOIDs to other SDOs (#2271)

bakkot committed:
Editorial: rename BindingInstantiation to ForDeclarationBindingInstantiation (#2271)

bakkot committed:
Editorial: consolidate Contains (#2271)

bakkot committed:
Editorial: consolidate ComputedPropertyContains (#2271)

bakkot committed:
Editorial: prefer multi-line productions when in groups (#2271)

bakkot committed:
Editorial: factor out Instantiate*FunctionExpression SDOs (#2271)

bakkot committed:
Editorial: consolidate IsIdentifierRef (#2271)

bakkot committed:
Editorial: consolidate NamedEvaluation (#2271)

bakkot committed:
Editorial: move HasName, IsFunctionDefinition, and IsAnonymousFunctionDefinition to Function Name Inference section (#2271)

bakkot committed:
Editorial: add note about LabelledEvaluation special cases (#2271)

bakkot committed:
Editorial: consolidate IsFunctionDefinition (#2271)

syg committed:
Editorial: List the [[Construct]]-requires-[[Call]] invariant in the list of invariants (#2213)

ljharb committed:
Editorial: add missing assertion to ApplyStringOrNumericBinaryOperator (#2282)

Fixes #2281

bakkot committed:
Editorial: split up IterationStatement (#2280)

bakkot committed:
Meta: bump ecmarkup to 6.0.0 (#2289)

jmdyck committed:
Editorial: Merge the RHSs of AsyncFunctionExpression (#2275)

The nonterminals:
- FunctionExpression
- ClassExpression
- GeneratorExpression
- AsyncGeneratorExpression

are all defined with a single RHS involving an optional BindingIdentifier.

But AsyncFunctionExpression is defined with two RHSs, one with a BindingIdentifier and one without.
(It's been that way since it was introduced in PR #692)
I can't see any reason for it to be not like the others.

mathiasbynens committed:
Editorial: improve “About this Specification” section (#2284)

Published snapshots such as https://262.ecma-international.org/11.0/ include the exact same text. This patch tweaks the text so that the sentence still makes sense even when it’s being read from a snapshot copy. It also hardcodes the link’s target URL so that the link text and destination are always in sync and point to the expected location.

DanielRosenwasser committed:
Editorial: swap phrasing to indefinite article for Boolean-coercibles (#2295)

ljharb committed:
Editorial: add <dfn> for "TypedArray element type" (#2297)

Commit:
Editorial: convert Number abstract operations to algorithmic steps (#2189)

Co-authored-by: Ryan Duffy <ryan.duffy@lge.com>
Co-authored-by: Michael Ficarra <mficarra@shapesecurity.com>
Co-authored-by: Kevin Gibbons <bakkot@gmail.com>

bakkot committed:
Meta: bump ecmarkup to 6.0.1 (#2301)

bakkot committed:
Editorial: replace a "<=" with a "≤" (#2301)

szuend committed:
Normative: A more precise Array.prototype.sort (#1585)

Co-authored-by: Simon Zünd <szuend@chromium.org>
Co-authored-by: Mathias Bynens <mathias@qiwi.be>

rkirsling committed:
Normative: Give %TypedArray% methods explicit algorithms (#2221)

ljharb committed:
Editorial: remove unnecessary assertion (#2296)

Both S and R are Strings, so either both or neither should be asserted in the steps. I chose neither.

ExE-Boss committed:
Normative: Set function `name` and `length` in `CreateBuiltinFunction` (#2116)

Co-authored-by: Kevin Gibbons <bakkot@gmail.com>
Co-authored-by: Shu-yu Guo <syg@chromium.org>

ExE-Boss committed:
Editorial: Improve callback variable names used by `Promise` statics (#2116)

jmdyck committed:
Editorial: Standardize preambles for host-defined abstract operations (#2253)

... using the format established in PR #1914.

ljharb committed:
Markup: add missing `</p>` tag from 346e583 / #2116

bakkot committed:
Meta: bump ecmarkup to 6.1.0 (#2311)

bakkot committed:
Editorial: dfn Normative Optional (#2311)

syg committed:
Editorial: Remove paragraph listing similarities to other languages (#2312)

ljharb committed:
Meta: add `scroll-margin-top` CSS to make for a nicer anchor experience (#2317)

See https://piccalil.li/quick-tip/add-scroll-margin-to-all-elements-which-can-be-targeted

ExE-Boss committed:
Editorial: Add support for Abstract Closures to `CreateBuiltinFunction` (#2109)

Co-authored-by: ExE Boss <3889017+ExE-Boss@users.noreply.github.com>
Co-authored-by: Kevin Gibbons <bakkot@gmail.com>
Co-authored-by: Michael Dyck <jmdyck@ibiblio.org>
Co-authored-by: Jordan Harband <ljharb@gmail.com>
Co-authored-by: Michael Ficarra <mficarra@shapesecurity.com>

ExE-Boss committed:
Editorial: Fix `<dfn>`s for `WeakRef` and `FinalizationRegistry` (#2318)

bakkot committed:
Normative: forbid for-of loops with variable named async (#2256)

ljharb committed:
Editorial: swap "about" and "contributing" sections in the intro (#2324)

ljharb committed:
Editorial: nest ASI sections under "ASI" section (#2323)

devsnek committed:
Normative: Define default constructors using spec steps (#2216)

shvaikalesh committed:
Editorial: Align checks order in [[DefineOwnProperty]] overrides with ordinary method (#2333)

Commit:
Editorial: Fix mistakes (#2331)

jmdyck committed:
Editorial: Change "that satisfies" to "such that" (#2330)

The spec has only one occurrence of:
    "For each ... that satisfies ..."
but six of:
    "For each ... such that ..."
so change the former to the latter.

jmdyck committed:
Editorial: Change "non-negative integer" to "integer" (#2330)

... in the only occurrence of "For each non-negative integer",
to match the 4 existing occurrences of "For each integer".

jmdyck committed:
Editorial: Drop "in original [key] insertion order" (#2330)

The form
    "For each [element] of [list], in original key insertion order"
(or "... original insertion order") is weird because it implies that
it's not enough to know the order of the elements in the List,
you also have to know the order that they were inserted into the List.
But in fact, that's not necessary: in every case where this form is used,
the insertion order is the *same* as the List order,
because insertions only happen via appending to the List.

Map.prototype.forEach:
  Iterates over _map_.[[MapData]],
  whose only inserts are appends, in Map.prototype.set.

Set.prototype.forEach:
  Iterates over _set_.[[SetData]],
  whose only inserts are appends, in Set.prototype.add.

TriggerPromiseReactions:
  Iterates over _reactions_, which is
  _promise_.[[PromiseFulfillReactions]] or
  _promise_.[[PromiseRejectReactions]],
  whose only inserts are appends, in PerformPromiseThen.

bakkot committed:
Meta: better compression during preview uploads (#2340)

ljharb committed:
Editorial: add ES2021 intro text

ljharb committed:
Editorial: master is now ES2022

ljharb committed:
Meta: update editors list for ES2022

mathiasbynens committed:
Editorial: Clarify ES2021 Array.prototype.sort change (#2342)

`Array.prototype.sort` was already made stable in ECMAScript 2019 (see two paragraphs above). The change in https://github.com/tc39/ecma262/pull/1585 made `Array.prototype.sort` more precise, reducing the amount of cases that result in an implementation-defined sort order.

bakkot committed:
Normative: specify creation order for capturing group properties (#2329)

bakkot committed:
Meta: bump ecmarkup to 7.0.2 (#2339)

bakkot committed:
Meta: multipage builds (#2339)

syg committed:
Normative: Use HostJobCallbacks for FinalizationRegistry callbacks (#2316)

This aligns FinalizationRegistry callbacks with Promise callbacks in
allowing web browsers to perform additional steps to save and restore
the correct incumbent settings object.

jmdyck committed:
Editorial: Have NewGlobalEnvironment invoke NewObjectEnvironment (#2287)

... rather than creating "a new object Environment Record" directly.

(Now, all object Environment Records are created via NewObjectEnvironment.)

jmdyck committed:
Editorial: Set _withEnvironment_ explicitly (#2287)

... when creating an object Environment Record.

(So _withEnvironment_ no longer needs to have a "default" value.)

jmdyck committed:
Editorial: Add parameter to NewObjectEnvironemt (#2287)

... that gives the value for the new environment's _withEnvironment_ flag.

jmdyck committed:
Editorial: Introduce fields to specify object Env Rec (#2287)

Specifically: [[BindingObject]] and [[IsWithEnvironment]]

jmdyck committed:
Editorial: Rename alias `_bindings_` to `_bindingObject_` (#2287)

michaelficarra committed:
Editorial: use target alias name instead of "it" (#2341)

bakkot committed:
Meta: bump ecmarkup to 7.0.3 (#2344)

mathiasbynens committed:
Editorial: remove redundant `index.html` (#2344)

This is a follow-up to #2339.

michaelficarra committed:
Editorial: clean up wording of Array and TypedArray methods (#2346)

jmdyck committed:
Editorial: 5 simple tweaks that I believe are self-explanatory (#2351)

jmdyck committed:
Editorial: Change "that is the same sign as _number_" (#2351)

to "whose sign is the sign of _number_"

(My first thought was to simply change "is" to "has":
`the mathematical value that *has* the same sign as _number_ ...`
but then "and whose magnitude ..." still sounded odd,
so instead I'm suggesting a "whose sign" clause to balance it.)

jmdyck committed:
Editorial: Drop Script Record's [[Environment]] field (#2351)

It's set to `*undefined*` in ParseScript and then never referenced again.
(Instead, we have _scriptRecord_.[[Realm]].[[GlobalEnv]])

jmdyck committed:
Editorial: Change "headings" to "heading" (#2351)

I believe the spec referred to "headings" in
   "subclause headings for the function description"
to accommodate a few functions that were defined
via 'overloads' in multiple clauses.

But PR #2176 rewrote those functions so that each has
a single definition, and thus only one relevant header.

jmdyck committed:
Editorial: Change "includes" to "contains" (#2351)

I think this is the only (remaining) use of "include" with this sense.

See https://github.com/tc39/ecma262/pull/1554#discussion_r577744595

h2oche committed:
Editorial: Fix early error rules of 'IfStatement' (#2359)

jmdyck committed:
Editorial: Unfold SDO `CoveredParenthesizedExpression` (#2350)

jmdyck committed:
Editorial: Unfold SDO `CoveredCallExpression` (#2350)

jmdyck committed:
Editorial: Unfold SDO `CoveredAsyncArrowHead` (#2350)

jmdyck committed:
Editorial: Eliminate one use of SDO `CoveredFormalsList` (#2350)

Here in InstantiateArrowFunctionExpression,
I assert that we can replace the status quo's
    CoveredFormalsList of |ArrowParameters|
with just
    |ArrowParameters|
(and then eliminate the _parameters_ alias).

case analysis on the 2 forms of |ArrowParameters|:

case 1:
When |ArrowParameters| is an instance of
    ArrowParameters : BindingIdentifier
then CoveredFormalsList simply returns |ArrowParameters| itself,
so the above replacement is obviously correct.

case 2:
When |ArrowParameters| is an instance of
    ArrowParameters : CoverParenthesizedExpressionAndArrowParameterList
then CoveredFormalsList will return an instance of
    ArrowFormalParameters
(whose particulars are not important here),
which then gets passed to OrdinaryFunctionCreate's ParameterList.

But if you go to OrdinaryFunctionCreate and follow what happens
to ParameterList, you find that, ultimately, it's submitted to:
    - ExpectedArgumentCount
    - BoundNames
    - IsSimpleParameterList
    - ContainsExpression
    - IteratorBindingInitialization

And all of these have an explicit definition for:
    ArrowParameters : CoverParenthesizedExpressionAndArrowParameterList
of roughly the form:
    1. Let _formals_ be the CoveredFormalsList of |CoverParenthesizedExpressionAndArrowParameterList|.
    1. Return <SDO> of _formals_.

That is, with respect to the OrdinaryFunctionCreate's call-tree,
if you don't call CoveredFormalsList near the root,
it will be called near the leaves.

So in this case too, it's okay to perform the given replacement.

jmdyck committed:
Editorial: Unfold SDO `CoveredFormalsList` (#2350)

h2oche committed:
Editorial: Consistent notation of newly created error object (#2363)

jmdyck committed:
Editorial: Factor out abstract operation RegExpHasFlag (#2349)

h2oche committed:
Editorial: Correct usage of 'Let'/'Set' in various algorithms (#2365)

jmdyck committed:
Editorial: Simplify CreateDynamicFunction (#2348)

Formerly, CreateDynamicFunction would parse the text for the parameters and body *separately*, so we didn't have a complete function decl/expr, so the Early Error rules for the function decl/expr production didn't apply automatically, so we had to apply them "manually".

Instead, parse the text of a complete function expr (which text already exists in _sourceText_!), so that all the relevant Early Error rules apply automatically.

h2oche committed:
Editorial: Fix Contains algorithm of ClassTail (#2362)

bakkot committed:
Editorial: suppress spurious linking of "Contains" (#2369)

bakkot committed:
Editorial: more explicit validation in CreateDynamicFunction (#2374)

jmdyck committed:
Editorial: Copy EscapeSequence to TemplateEscapeSequence (#1867)

... and replace the former with the latter in TemplateCharacter.

The spec prohibits applying the B.1.2 extensions
to EscapeSequence when it appears in TemplateCharacter;
this change accomplishes that in the grammar rather than in prose.

(This makes it a little easier to 'mainline' the B.1.2 extensions
in the next commit.)

Ms2ger committed:
Normative: Add public and private class fields, static class features, private methods & getters/setters

This includes changes from https://tc39.es/proposal-class-fields/.
This includes changes from https://tc39.es/proposal-private-methods/.
This includes changes from https://tc39.es/proposal-static-class-features/.

Co-authored-by: Ms2ger <Ms2ger@gmail.com>
Co-authored-by: Ujjwal Sharma <ryzokuken@igalia.com>
Co-authored-by: Caio Lima <ticaiolima@gmail.com>
Co-authored-by: Michael Dyck <jmdyck@ibiblio.org>
Co-authored-by: Kevin Gibbons <bakkot@gmail.com>
Co-authored-by: Jordan Harband <ljharb@gmail.com>

syg committed:
Normative: Make SharedArrayBuffer's sole parameter required (#2393)

SharedArrayBuffer.length is 1 on all web engines (as is
ArrayBuffer.length). Unsure why this was made optional.

eshost output for `print(SharedArrayBuffer.length)`:

```
1

1

1

1
```

bakkot committed:
Editorial: consistency for capitalization in notes (#2419)

bakkot committed:
Meta: bump ecmarkup to 7.1.0 (#2419)

jmdyck committed:
Editorial: Change ValidateTypedArray so that it doesn't return the buffer (#2376)

See https://github.com/tc39/ecma262/pull/2355#issuecomment-815296498

angleKH committed:
Editorial: fix typo in Iterator Interface "next" requirements (#2424)

himsngh committed:
Editorial: `meta` should be listed alongside target (#2416)

Fixes #2298

bakkot committed:
Editorial: fix handling of infinities in Math.hypot (#2426)

bakkot committed:
Editorial: consistency for FlattenIntoArray's optional args (#2427)

bakkot committed:
Editorial: drop remaining uses of Concrete Method suffix (#2427)

There's lots of other concrete methods, but only the module ones had the literal "Concrete Method" suffix in their header.

bakkot committed:
Editorial: algorithms constraining memory model are not AOs (#2427)

bakkot committed:
Editorial: move *Instantiation notes to after their preambles (#2427)

This is the usual convention, and these are the only places in the current spec where an AO's header is separated from its preamble.

bakkot committed:
Editorial: make preamble for LocalTZA consistent with other AOs (#2427)

bakkot committed:
Editorial: consistency for preambles Create*Iterator AOs (#2427)

bakkot committed:
Editorial: put WaiterList in its own clause (#2427)

bakkot committed:
Editorial: emu-notes in AOs go before "It performs..." (#2427)

bakkot committed:
Editorial: fix note about var decls in eval in parameter initializers (#2428)

jmdyck committed:
Editorial: Recast "Abstract Relational Comparison" (#2378)

... as a proper abstract operation, named IsLessThan.

jmdyck committed:
Editorial: Recast "Abstract Equality Comparison" (#2378)

... as a proper abstract operation, named IsLooselyEqual.

jmdyck committed:
Editorial: Recast "Strict Equality Comparison" (#2378)

... as a proper abstract operation, named IsStrictlyEqual.

jmdyck committed:
Editorial: Make IsLessThan's _LeftFirst_ param required (#2378)

tomayac committed:
Meta: CONTRIBUTING.md: Some changes to make the doc more beginner-friendly (#2431)

 - Add new line after headings
 - Link the discourse
 - Minor word changes

ljharb committed:
Meta: migrate from travis-ci to Github Actions (#2260)

bakkot committed:
Editorial: use abstract closures in Await (#2439)

bakkot committed:
Editorial: use abstract closure for default constructor (#2439)

bakkot committed:
Editorial: use abstract closure in AsyncFromSyncIteratorContinuation (#2439)

bakkot committed:
Editorial: use abstract closure in NewPromiseCapability (#2439)

bakkot committed:
Editorial: use abstract closure in Promise.prototype.finally (#2439)

bakkot committed:
Editorial: use abstract closures in AsyncGeneratorResumeNext (#2439)

bakkot committed:
Editorial: use abstract closure in Proxy.revocable (#2439)

bakkot committed:
Editorial: use abstract closures in MakeArg{Getter,Setter} (#2439)

ljharb committed:
Meta: add debug info (see #2260)

This is pushed directly to the default branch, because workflow_run scripts are ran from the default branch

ljharb committed:
Meta: fix passing PR number (see #2260)

ljharb committed:
Normative: add `#x in obj` syntax (#2290)

See https://github.com/tc39/proposal-private-fields-in-in/issues/7

bakkot committed:
Meta: bump ecmarkup to 8.0.0 (#2452)

bakkot committed:
Editorial: have Create{,Async}IteratorFromClosure provide a new execution context to GeneratorStart (#2398)

bakkot committed:
Editorial: Clarify the AsyncGenerator state machine (#2413)

bakkot committed:
Meta: fix auto-deployment (#2453)

jmdyck committed:
Editorial: Define+use "list-concatenation" (#2383)

michaelficarra committed:
Editorial: refactor runtime MV and add/use RoundStringMVResult (#2435)

reword some things that read like initialisms

fix linting error

accept some changes

extract StringNumericValue from MV on the string grammar

use StringNumericValue

further describe "the mathematical value denoted by"

address review comments

don't round NonDecimalIntegerLiteral

jmdyck committed:
Editorial: Drop grammatical parameters from 'use' productions (#2446)

... since they aren't mentioned by the associated algorithm/rule.

bakkot committed:
Editorial: fix typo in Number.prototype.toExponential (#2450)

bakkot committed:
Editorial: comparisons in number stringification are of floats, not reals (#2450)

bakkot committed:
Editorial: restore exponent in NumericValue (#2450)

h2oche committed:
Editorial: Fix wrong assertion in IsLessThan (#2456)

jmdyck committed:
Markup: Introduce structured headers (#545)

for:
 - abstract operations
 - numeric methods
 - 'concrete' methods
 - internal methods

ptomato committed:
Editorial: Remove 'designed to be subclassable' (#2360)

As discussed during the March 2021 plenary, this note on "is designed to
be subclassable" can be read as an encouragement or value judgement on
whether it is desirable to subclass built-in classes like Boolean. We
don't want to actively encourage this.

ljharb committed:
Meta: try alternate yaml syntax in preview-build workflow

syg committed:
Normative: Restrict completion values callables can return (#2448)

jmdyck committed:
Editorial: Inline "Dynamic Function SourceText Prefixes" table (#2367)

... into the if-else cascade in CreateDynamicFunction, suggested in https://github.com/tc39/ecma262/pull/2348#discussion_r606347643

jmdyck committed:
Editorial: Remove ToNumber's note re surrogates (#1554)

The note says that "the result of ToNumber will be *NaN*
if the string contains any [surrogate] code units,
whether paired or unpaired."

However, if Unicode were to define a non-BMP code point
with General_Category=Zs, that would qualify as <USP>
and thus WhiteSpace and StrWhiteSpaceChar, in which case
the note would be rendered false. That is, the ToNumber
procedure would continue to work, and would result in
non-NaN values for some strings containing surrogates.

(And if you think that the ES spec doesn't need to concern itself
with such future possibilities, note that ES 6 (2015) changed/clarified
the semantics of String.p.trim et al for precisely this case,
to say how they would deal with non-BMP white space.)

Given that the note could be invalidated by a future
edition of Unicode, I think it's a bit risky to keep it in.

jmdyck committed:
Editorial: Define + use StringToNumber (#1554)

Introduce the abstract operation StringToNumber
to replace the prose that expressed the procedure
for applying ToNumber to String values.

pacokwon committed:
Editorial: Validate array length using `SameValueZero` (#2460)

syg committed:
Normative: Constrain all host hooks to return either normal or throw completions (#2442)

This PR also makes the prose around requirements for host hook
implementations consistent to the template "An implementation of
HostHook must conform to the following requirements:".

bakkot committed:
Meta: bump ecmarkup to 8.0.1 (#2467)

ljharb committed:
Editorial: add "Legacy" section (#2125)

 - co-opts prose from Annex B description
 - dfn legacy and add emu-not-ref where appropriate
 - use dash in IDs
 - don't <dfn> legacy; clauses are what is legacy

Co-authored-by: Jordan Harband <ljharb@gmail.com>
Co-authored-by: Michael Ficarra <mficarra@shapesecurity.com>

devsnek committed:
Normative: move __proto__ out of annex b (#2125)

Co-authored-by: Gus Caplan <me@gus.host>
Co-authored-by: Jordan Harband <ljharb@gmail.com>
Co-authored-by: Michael Dyck <jmdyck@ibiblio.org>

michaelficarra committed:
Editorial: don't imply that SV, TV, and TRV are shorthands (#2451)

devsnek committed:
Editorial: initialize missing fields in OrdinaryFunctionCreate (#2464)

bakkot committed:
Markup: prefer `&notin;` over `&lt;!` (#2469)

ExE-Boss committed:
Editorial: Fix `internalSlotsList` parameter of `CreateBuiltinFunction` (#2115)

jmdyck committed:
Editorial: Split the "TV and TRV" clause into 2 clauses (#2375)

... one for TV and one for TRV.

(This duplicates 4 cases, but they're very simple.)

bakkot committed:
Editorial: drop constant arguments for PropertyDefinitionEvaluation and ClassElementEvaluation (#2470)

bakkot committed:
Editorial: consolidate method cases of PropertyDefinitionEvaluation (#2470)

bakkot committed:
Editorial: avoid duplicate SetFunctionName in {async,}generator methods (#2463)

gibson042 committed:
Editorial: Fix alias use in IteratorBindingInitialization (#2477)

codehag committed:
Normative: Allow null or undefined in Reference Records (#2267)

Co-authored-by: codehag <yulia.startsev@gmail.com>
Co-authored-by: Kevin Gibbons <bakkot@gmail.com>
